doc/port_fec-ber/port_fec_ber.md

```diff
@@ -0,0 +1,205 @@
+
+|Redis DB |Table|Entries|New, RW| Format | Descriptions|   
+|--------------|-------------|------------------|--------|----------------|----------------|  
+|COUNTER_DB |COUNTERS |SAI_PORT_STAT_IF_IN_FEC_CORRECTED_BITS |R |number |Total number bits corrected</sub>|
+|COUNTER_DB |COUNTERS |SAI_PORT_STAT_IF_IN_FEC_NOT_CORRECTABLE_FRAMES |R |number |Tota number uncorrectable frame |
+|COUNTER_DB |COUNTERS_PORT_NAME_MAP |SAI_PORT_STAT_IF_IN_FEC_NOT_CORRECTABLE_FRAMES|R |number |Oid to name mapping |  
+|COUNTER_DB |RATES |SAI_PORT_STAT_IF_IN_FEC_CORRECTED_BITS_last | New, RW|number |Last corrected bits counts |
+|COUNTER_DB |RATES |SAI_PORT_STAT_IF_IN_FEC_NOT_CORRECTABLE_FRAMES_last |New, RW|number |Last uncorrctedble frame counts |  
+|COUNTER_DB |RATES |FEC_PRE_BER |New, RW| floating |calculated pre fec BER |  
+|COUNTER_DB |RATES |FEC_POST_BER |New, RW| floating | calulated post fec BER |  
+|APPL_DB |PORT_TABLE |lanes |R |strings |number of serdes lanes in the port |  
+|APPL_DB |PORT_TABLE |speed |R |number |port speed |  
+
+
+### 4.3 SAI API
+
+No change in the SAI API. No new SAI object accessed.
+
+
+### 4.4 Calculation Formulas
+
+Each port can be made up of multiple lanes and each running at the same serdes speed. The hardware counter is per port basis. Therefore the BER calculation will require to account for #lanes and serdes speed of the port.  The lanes count and port speed can be retrieved from the APPL_DB. Serdes speed will be calculated using port speed / number of lanes.
+
+However in the case of post FEC, the frames are actually dropped. There is no way to tell the actual number of error bits in the frame. The calculation is assuming the worst case where all bits in the frame were corrupted.
+
+
+
+```
+Pseudocode :
+
+Step 1: get lanes count and port speed
+    lanes count  : APPL_DB , hget PORT_TABLE lanes , lane count calculate using string.gsub()
+    port speed : APPL_DB , hget PORT_TABLE speed,  speed of the port
+
+Step 2: calculate port_data_rate
+    port_data_rate = math.fmod(port speed / number of lanes)
+
+Step 3 : look up link serdes speed
+    Look up the serdes speed using the following logic
+
+    if port_data_rate == 1000 then
+        serdes = 1.25e+9
+    elseif port_data_rate == 10000 then
+        serdes = 10.3125e+9
+    elseif port_data_rate == 25000 then
+        serdes = 25.78125e+9
+    elseif port_data_rate == 50000 then
+        serdes = 53.125e+9
+    elseif port_data_rate == 100000 then
+        serdes = 106.25e+9
+    else
+        serdes = 0
+    end
+
+Step 4 : frame size
+
+     /* NRZ encoding Rs-528  or PAM4 encoding Rs-544*/
+    if user_port_lane_speed <= 25G
+      frame_size = 5280
+    else
+      frame_size = 5440
+
+Step 5: calcuate BER
+
+    interval = PORT_STATE poll interval which is 1000 ms currently
+
+    Pre FEC BER = (SAI_PORT_STAT_IF_IN_FEC_CORRECTED_BITS - SAI_PORT_STAT_IF_IN_FEC_CORRECTED_BITS_last) / (serdes * lanes_count * interval / 1000)
```

Comment:
why is the division by 1000 at the end of the calculation ?
BER in interval = (# errors / # bits) in interval
#bits = serdes rate * lane * interval

Reply:
The polling is done through the port_rates.lua script. The script were run at 1000ms interval as default. The interval were pass to the script as an argument which is in ms... that is 1 sec were pass in as 1000.
Now in order to account for the total number of bit ingress.. we need to multiple the serdes bps  with the total duration is sec, hence  the reason for divided by 1000

Reply:
I agree that the bits passed are multiplied by the interval number.
My lack of understanding is why we have both multiplication of interval variable and division of 1000.
the total amount of bits passed in 1000ms interval are: serdes * lanes_count * interval
Dividing by another 1000 is redundant no?.

Reply:
it is because the interval which were passing is equal to 1000,  and can be changed with a cli as needed and the value it took is in ms. Hence when we multiple by interval we need to divide it by 1000.